File: exercism/javascript/luhn/luhn.js
const fp = require('lodash/fp')

const mapWithIndex = fp.map.convert({cap: false})

const digits = (n) =>
  String(n).split('').map(fp.parseInt(10))

const trans = (d, i) =>
  i % 2 == 0 ? d : (d < 5 ? d * 2 : d * 2 - 9)

class Luhn {
  constructor (n) {
    this.checkDigit = n % 10
    this.addends = fp.reverse(mapWithIndex(trans, fp.reverse(digits(n))))
    this.checksum = fp.sum(this.addends)
    this.valid = this.checksum % 10 == 0
  }
}

Luhn.create = (n) => {
  const candidate = n * 10
  const checksum = new Luhn(candidate).checksum % 10
  return candidate + (10 - checksum) % 10
}

module.exports = Luhn

File: exercism/csharp/luhn/Luhn.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Luhn
{
    public long CheckDigit;
    public IEnumerable<int> Addends;
    public int Checksum;
    public bool Valid;

    public Luhn (long given)
    {
        CheckDigit = given % 10;
        Addends = given.ToString()
            .Reverse()
            .Select((ch, i) => i % 2 == 0 ? (ch - '0') : Wrap(ch))
            .Reverse();
        Checksum = Addends.Sum();
        Valid = Checksum % 10 == 0;
    }

    public static long Create (long given)
    {
        var candidate = given * 10;
        var l = new Luhn(candidate);
        if (l.Valid) return candidate;
        return candidate + 10 - l.Checksum % 10;
    }

    private int Wrap (char ch)
    {
        var n = (ch - '0') * 2;
        return n > 9 ? n - 9 : n;
    }

}
